### PR TITLE
Group using Hash and Sort

### LINKED ISSUE #12: Group using hash and sort
The current implementation relies on CLI `sort`.  Used Python's sort instead.

EDIT: it would be even better to improve the group strategy to use `hash(key) % num_reducers`

### PR DESCRIPTION
**Pull request by awdeorio**
Modify the group stage to allocate lines to group by hash of key.  Then, sort each group.  This matches the Apache Hadoop Streaming interface.

Closes #12

### CODE REVIEW

**Comment by jaredzh** [comment]
Should I just ignore the failed checks from codecov? @awdeorio

**Comment by awdeorio** [comment]
> Should I just ignore the failed checks from codecov? @awdeorio

Yea, that's fine for now

**Comment by awdeorio** [comment]
Bump @jaredzh !

**Comment by jaredzh** [comment]
I left a comment earlier, did you get a chance to look at it? @awdeorio

**Comment by awdeorio** [comment]
> I left a comment earlier, did you get a chance to look at it?

The only comment I saw was a question about code cov, which I responded to.  Maybe I missed somethign?

**Comment by jaredzh** [comment]
Hmm interesting, you should be able to see them under the "files changed" tab. Anyways, they were basically addressing the nitpicks I pushed. Take a look and go ahead and merge if it's good.

**Review by jaredzh — APPROVED** [review]

File: madoop/mapreduce.py [hunk 1]
```diff
@@ -176,79 +176,61 @@ def map_stage(exe, input_dir, output_dir):
...
-            grouper_files[0].write(line)
+    # Sort output files
+    for path in output_dir.iterdir():
+        sort_file(path)
+
+    # Remove empty output files.  We won't always use the maximum number of
+    # reducers because some MapReduce programs have fewer intermediate keys.
+    for path in output_dir.iterdir():
+        if path.stat().st_size == 0:
+            path.unlink()
```
[the remark this answers is not in the document]
Comment (awdeorio) [reply madoop/mapreduce.py:224]: Yep!  Thanks for actually reading the code!

File: tests/test_stages.py [hunk 2]
```diff
@@ -19,6 +20,7 @@ def test_map_stage(tmpdir):
 
 def test_group_stage(tmpdir):
     """Test group stage using word count example."""
+    tmpdir = Path(tmpdir)
```
[the remark this answers is not in the document]
Comment (awdeorio) [reply tests/test_stages.py]: All three stages assume that the inputs are `PosixPath` objects, not strings.  The only time this is a problem is when we "white box test" the stages here.

For reference, `tmpdir` is (confusingly) not a `PosixPath`.  It's also not a `str`!

**Pull request merged**